### DISCUSSION TITLE
Recommended way for client code that connects to many servers.

### DISCUSSION THREAD

**Discussion by schnabel** [miguelgrinberg/python-socketio#832] [Q&A] [answered]
Hi, 
I am writing a client api which could be used to connect to multiple server instances. I found a few discussions about the topic:

1. [How use socketio decorator within another class](https://github.com/miguelgrinberg/python-socketio/issues/390) 
2. [Python-socketio: How to connect one client to multiple servers?](https://stackoverflow.com/questions/67595720/python-socketio-how-to-connect-one-client-to-multiple-servers)
3. [How do you prevent an instance using a socketio.AsyncClient code from blocking?](https://stackoverflow.com/questions/68973680/how-do-you-prevent-an-instance-using-a-socketio-asyncclient-code-from-blocking)

Is there a recommended 'class based' approach? Is there a way to use the decorators with instance methods? At the moment I use a combination of solution 3 and Class-Based Namespaces.

**Comment by miguelgrinberg** [comment, marked as answer]
@schnabel Is the question how you write a wrapper class that combines the Socket.IO client instance and the event handlers all in the same class? I'm not sure I would do that myself, so I have no recommendation to make. But It is possible in Python to do this. You could add the client instance and the connection/disconnection methods to your namespace subclass, for example.

**Reply by schnabel** [reply]
Hi Miguel, 
here is a little snippet:
```
class ExampleController:

    def __init__(self, host: str = 'localhost', port: int = 8080) -> None:
        self.host = host
        self.port = port
        self.callbacks = []
        self.sio = socketio.AsyncClient()
        self.example_namespace = ExampleNamespace(self.sio)

    async def connect(self):
        self.sio.register_namespace(self.example_namespace)
        await self.sio.connect(f'http://{self.host}:{self.port}', wait_timeout=10)

class ExampleController(socketio.AsyncClientNamespace):

    def __init__(self, sio):
        super().__init__(namespace="/example")
        self.sio = sio

    async def hello(self):
        await self.sio.emit("hello", namespace="/example")
```
My question is, if this is not a recommended way to achieve connections to many different servers, how would you do it :-).

**Reply by miguelgrinberg** [reply]
I can't really answer. You are asking me to recommend an approach but you are putting some constraints, such as that there should be an encapsulating class that includes the Socket.IO client instance and the handlers, and I already said that I wouldn't do this encapsulation myself. So I really have nothing that I feel deserves my recommendation, given those constaints.

But you are assuming that if I don't recommend something, then it is automatically a "negative" recommendation, and this isn't the case. If you feel it is important to encapsulate the client and the handlers in a single class, then you are free to do it. I'm not saying it is wrong, just that it wouldn't be my approach.

**Reply by schnabel** [reply]
Oh, my snippet was not meant to show any constraints. I only wanted to show the way I tried to implement it. 
I am just curious how others (and especially you) would address connecting to more than one server from the same client.

**Reply by miguelgrinberg** [reply]
You can't connect to multiple servers with the same client instance. You need a different instance per connection.

```python
client1 = Client()
client1.register_namespace(MyNamespace())
client1.connect('https://server1.com')

client2 = Client()
client2.register_namespace(MyNamespace())
client2.connect('https://server2.com')
```

**Reply by schnabel** [reply]
Again I should have made myself clearer, sorry. I meant 'client' application not socketio.Client :-). 
Your example is great, if you know how many servers there are. Unfortunately this is not the case for me :-(.

**Reply by miguelgrinberg** [reply]
```python
def make_socketio_client(url):
    client = Client()
    client.register_namespace(MyNamespace())
    client.connect(url)
    return client
```

**Reply by schnabel** [reply]
This seems like a way to go for me. Keep it simple :-). Thanks a lot for your suggestions and of course for python-socketio. Great job.
